tachyon/zk/plonk/permutation/permutation_assembly.h

```diff
@@ -0,0 +1,245 @@
+    }
+
+    Label tmp = mapping_[left_col_idx][left_row];
+    mapping_[left_col_idx][left_row] = mapping_[right_col_idx][right_row];
+    mapping_[right_col_idx][right_row] = tmp;
+
+    return true;
+  }
+
+  // Returns `PermutationVerifyingKey` which has commitments for permutations.
+  constexpr PermutationVerifyingKey<Curve> BuildVerifyingKey(
+      EvaluationDomainTy* domain) const {
+    std::vector<EvalsPolyTy> permutations = GeneratePermutations(domain);
+
+    // TODO(dongchangYoo): calculate commitments after complete Params
+    std::vector<CommitmentTy> commitments;
+
+    return PermutationVerifyingKey<Curve>(commitments);
+  }
+
+  // Returns the `PermutationProvingKey` that has the coefficient form and
+  // evaluation form of the permutation.
+  constexpr PermutationProvingKey<FieldTy, MaxDegree> BuildProvingKey(
+      EvaluationDomainTy* domain) const {
+    // The polynomials of permutations in evaluation form.
+    std::vector<EvalsPolyTy> permutations = GeneratePermutations(domain);
+
+    // The polynomials of permutations with coefficients.
+    std::vector<CoeffsPolyTy> polys;
+    for (size_t i = 0; i < columns_.size(); ++i) {
+      CoeffsPolyTy poly = domain->IFFT(permutations[i]);
+      polys.push_back(poly);
+    }
+
+    return PermutationProvingKey<FieldTy, MaxDegree>(permutations, polys);
+  }
+
+  // Calculate 𝛿 = g^2ˢ with order "kTrace". (i.e., {kTrace}-th root of unity)
+  constexpr FieldTy GetDelta() const {
+    FieldTy g = FieldTy(FieldTy::Config::kSubgroupGenerator);
+    typename FieldTy::BigIntTy s(FieldTy::Config::kTwoAdicity);
+
+    FieldTy two(2);
+    FieldTy adicity = two.Pow(s);
+    return g.Pow(adicity.ToBigInt());
+  }
+
+  // Generate a 2 dimensional table which has label consist of 𝛿 and w.
+  constexpr std::vector<std::vector<FieldTy>> ConstructLabelTable(
+      EvaluationDomainTy* domain) const {
+    // The omega is gᵀ with order 2ˢ where modulus = 2ˢ * T + 1.
+    FieldTy omega(FieldTy::Config::kTwoAdicRootOfUnity);
+    std::vector<FieldTy> omega_powers = domain->GetRootsOfUnity(N, omega);
+
+    // The 𝛿 has order T where modulus = 2ˢ * T + 1.
+    FieldTy delta = GetDelta();
+
+    std::vector<std::vector<FieldTy>> label_table;
+    // Assign [𝛿⁰w⁰, 𝛿⁰w¹, 𝛿⁰w², ..., 𝛿⁰wⁿ⁻¹] to the first row.
+    label_table.push_back(omega_powers);
+
+    // Assign [𝛿ⁱw⁰, 𝛿ⁱw¹, 𝛿ⁱw², ..., 𝛿ⁱwⁿ⁻¹] to each row.
+    FieldTy delta_iter = delta;
+    for (size_t i = 1; i < columns_.size(); ++i) {
+      std::vector<FieldTy> label_row = base::CreateVector(N, FieldTy::Zero());
+      for (size_t j = 0; j < N; ++j) {
+        label_row[j] = omega_powers[j] * delta_iter;
+      }
+      label_table.push_back(label_row);
+      delta_iter *= delta;
+    }
+    return label_table;
+  }
+
+  // Generate the permutation polynomials based on the accumulated copy
+  // permutations. Note that the permutation polynomials are in evaluation form.
+  constexpr std::vector<EvalsPolyTy> GeneratePermutations(
```

Comment:
Can't be run at compile time. It's multithreaded.

```suggestion
  std::vector<EvalsPolyTy> GeneratePermutations(
```